byzcoin/statetrie.go

```diff
@@ -227,3 +238,18 @@ func newMemStateTrie(nonce []byte) (*stateTrie, error) {
 	}
 	return &st, nil
 }
+
+type metadataReader interface {
+	GetMetadata([]byte) []byte
+}
+
+func readVersion(t metadataReader) Version {
```

Comment:
I'm not sure this is correct. At what point is a new version stored in the block/stateTrie? Given only version-0 blocks, what happens if the following transactions get processed in block n:

1. BEVM-update with version 0
2. Config-update to version 1
3. BEVM-update with version **x**

What is **x** here? Would it be version 0 or version 1? Can you make a test that has an instruction that behaves differently for version 0 and version 1, and then create a block with the above clienttransaction list?

Also, did you go through what happens if a node wants to catch up?

What happens if a node wants to re-create a statechange-cache?

Reply:
I updated the test to illustrate how I was thinking the version would switch to a new one. Basically a block will be used to mark the beginning of a new version. Note that the function to create this block is not the definitive one but we still need to discuss with the others how we will make the switch so I just wrote it to show you.